File: src/lib/src/main/java/work/lclpnet/ap2/impl/util/GameRuleBuilder.java
package work.lclpnet.ap2.impl.util;

import net.minecraft.server.MinecraftServer;
import net.minecraft.world.level.gamerules.GameRule;
import net.minecraft.world.level.gamerules.GameRules;

public class GameRuleBuilder {

    private final GameRules gameRules;
    private final MinecraftServer server;

    public GameRuleBuilder(GameRules gameRules, MinecraftServer server) {
        this.gameRules = gameRules;
        this.server = server;
    }

    /**
     * Sets a boolean game rule value.
     * @param key A boolean game rule.
     * @param value The rule boolean value.
     * @return This builder instance.
     */
    public GameRuleBuilder set(GameRule<Boolean> key, boolean value) {
        gameRules.set(key, value, server);
        return this;
    }

    /**
     * Sets an integer game rule value.
     * @param key An integer game rule.
     * @param value The rule integer value.
     * @return This builder instance.
     */
    public GameRuleBuilder set(GameRule<Integer> key, int value) {
        gameRules.set(key, value, server);
        return this;
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/util/GameRuleBuilder.kt
package work.lclpnet.ap2.util

import net.minecraft.server.MinecraftServer
import net.minecraft.world.level.gamerules.GameRule
import net.minecraft.world.level.gamerules.GameRules
import work.lclpnet.ap2.game.MiniGameInstance

class GameRuleBuilder(
    private val gameRules: GameRules,
    private val server: MinecraftServer
) {

    /**
     * Sets a game rule value.
     * @param key The game rule.
     * @param value The rule value.
     * @return This builder instance.
     */
    fun <T : Any> set(key: GameRule<T>, value: T): GameRuleBuilder {
        gameRules.set(key, value, server)
        return this
    }
}

fun MiniGameInstance.useGameRules(config: GameRuleBuilder.() -> Unit) {
    config(GameRuleBuilder(level.gameRules, level.server))
}